Impl improved DocumentTable

ivy-edp opened the pull request:
In this PR, you can see my attempt to incorporate the following functionality (as discussed) into the DocumentTable. I would appreciate feedback:

- Header Redesign: Now resizable for column adjustment (reset with a double-click or component reload).
- Add-Row Button Redesign: Moved below as an Add-Line, no longer within the table (only possible to add one empty line, only visible when no empty row is at the bottom of the table).
- Row Interaction Buttons Redesign: Specific buttons for row interactions appear at the top right when a row is selected (also Delete-Row-Button is placed there).
- Keyboard Interaction Redesign: Tab through rows, and if the last row is not empty it adds a new row and after pressing tap again you should automaticelly be inside this new row.
- Cleanup Table: As soon as you deselect a row by clicking once on the header, the table removes all empty rows from the table. Afterwards, the Add-Row Button appears again.
![beispiel_DocumentTable_2](https://github.com/axonivy/inscription-client/assets/141223521/a6d41cb5-260c-4f7c-afa0-aa955609ab6c)

ivy-lli commented on packages/editor/src/components/widgets/table/cell/TableCell.tsx:
```diff
@@ -1,4 +1,8 @@
 import type { ReactNode } from 'react';
 
-export const TableCell = (props: { children: ReactNode }) => <td className='table-cell'>{props.children}</td>;
+export const TableCell = (props: { children: ReactNode; cellWidth?: number }) => (
+  <td className='table-cell' style={props.cellWidth ? { width: `${props.cellWidth}px` } : {}}>
+    {props.children}
+  </td>
+);
```
I would export all td properties and set this with on the TableCell directly...

```ts
export const TableCell = ({children, props}: React.HTMLAttributes<HTMLTableCellElement>) => (
  <td {...props} className='table-cell'>
    {children}
  </td>
);
```

```ts
<TableCell key={cell.id} style={{width: cell.column.getSize()}}>
  {flexRender(cell.column.columnDef.cell, cell.getContext())}
</TableCell>
```

ivy-edp replied:
when i do it like you suggest i get the error: Property 'props' does not exist on type 'HTMLAttributes<HTMLTableCellElement>'

isn't it maybe like this?
`export const TableCell = ({ children, ...props }: React.HTMLAttributes<HTMLTableCellElement>) => (
  <td {...props} className='table-cell'>
    {children}
  </td>
);`

ivy-lli replied:
yes, didn't wrote the code in an ide only here. so probably wrong. Only the idea ;)

ivy-lli commented on packages/editor/src/components/widgets/table/footer/TableFooter.tsx:
```diff
@@ -1,21 +1,31 @@
+        onClick={props.addRow}
+        disabled={readonly}
+        aria-label='Add row'
+        onFocus={() => setIsButtonHovered(true)}
+        onBlur={() => setIsButtonHovered(false)}
+        onMouseEnter={() => setIsButtonHovered(true)}
+        onMouseLeave={() => setIsButtonHovered(false)}
```
why do something via js what css can do too?

ivy-lli replied:
e.g.

```css
.add-row-container .button:hover ~ .add-line {
    background-color: var(--A300);
}
```

ivy-edp replied:
Oh my God, I've been looking for this specific functionality in CSS, but couldn't find it😂

ivy-lli commented on packages/editor/src/components/widgets/table/footer/TableFooter.css:
```diff
@@ -1,22 +1,51 @@
+  border-radius: 10px;
+}
+
+.add-row-container .button:hover {
+  background-color: var(--A300);
+  color: var(--background);
+}
+
+.add-row-container .button:focus {
+  background-color: var(--A300);
+  color: var(--background);
+}
```
same styles...

```css
.add-row-container .button:where(:hover, :focus) {
  background-color: var(--A300);
  color: var(--background);
}
```

ivy-lli commented on packages/editor/src/components/widgets/table/header/TableHeader.tsx:
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
 
-export const TableHeader = (props: { colSpan: number; children?: ReactNode }) => (
-  <th className='table-column-header' colSpan={props.colSpan}>
+export const TableHeader = (props: { colSpan: number; children?: ReactNode; columnWidth?: number }) => (
+  <th className='table-column-header' colSpan={props.colSpan} style={props.columnWidth ? { width: `${props.columnWidth}px` } : {}}>
     {props.children}
   </th>
 );
```
same here as on the TableCell

ivy-lli commented on packages/editor/src/components/widgets/table/header/TableHeader.tsx:
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
+    <>
+      {props.seperator && (
+        <div
+          {...{
+            onMouseDown: header.header.getResizeHandler(),
+            onTouchStart: header.header.getResizeHandler(),
+            className: `resizer ${header.column.getIsResizing() ? 'isResizing' : ''}`
+          }}
```
why wrap it in an object and then "unpack" it again?
is the following not working? 
```ts
<div onMouseDown={header.header.getResizeHandler()} onTouchStart={header.header.getResizeHandler()} className={`resizer ${header.column.getIsResizing() ? 'isResizing' : ''}`} />
```

ivy-lli commented on packages/editor/src/components/widgets/table/row/SelectRow.css:
```diff
@@ -1,14 +1,17 @@
   cursor: pointer;
 }
-.selectable-row:where(:hover) {
+.selectable-row:where(:hover),
+.selectable-row:where(:hover) .input {
```
where is not needed if there is only one thing inside, just do `.selectable-row:hover`

ivy-lli commented on packages/editor/src/components/widgets/table/row/SelectRow.css:
```diff
@@ -1,14 +1,17 @@
 }
 
 .selectable-row:where([data-state='selected']),
+.selectable-row:where([data-state='selected']) .table-cell .input,
```
same here, just do `.selectable-row[data-state='selected']`

ivy-lli commented on packages/editor/src/components/widgets/table/header/TableHeader.tsx:
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
-      />
-    </div>
+    <>
+      {props.seperator && (
```
a little bit strang that this resizer is the first element in this header and only because of some css it is on the "last" position...
Why isn't this just another item in the flex container? of course then you have to do something else than just `space-between` but that should definitly possible IMO...

ivy-lli commented on packages/editor/src/components/widgets/table/header/TableHeader.css:
```diff
@@ -29,3 +39,27 @@
+  user-select: none;
+  touch-action: none;
+}
+.table-column-header .resizer:hover {
+  width: 5px;
+  background: var(--A300);
+}
+
+.table-column-header .resizer.isResizing {
+  background: var(--A300);
+  width: 5px;
+  opacity: 1;
+}
```
why is opacity needed?
if this is gone you can combile this styles...

```css
.table-column-header .resizer:where(:hover, .isResizing) {
  width: 5px;
  background: var(--A300);
}
```

ivy-lli commented on packages/editor/src/components/widgets/headlineControls/HeadlineControls.css:
currently this is not used right? So I would revert this change!

ivy-lli commented on packages/editor/src/components/widgets/headlineControls/HeadlineControls.tsx:
currently this is not used right? So I would revert this change!

ivy-lli commented on packages/editor/src/components/widgets/fieldset/fieldset-control.ts:
currently this is not used right? So I would revert this change!

ivy-lli reviewed (changes requested):
Definitely a cool change!

However, I have the feeling that you do not check your changes well enough before adding a reviewer (or at least on this PR)...
I'm aware that you wanted feedback, but I'm a lot faster at review code if its is already cleaned up, e.g. from unused dead prototype code.

ivy-lli commented:
I'm confused, didn't you want to only do the document table for now? Why did you need also to change the CustomField table? 
If you made a copy of the components instead of changing the existing one you wouldn't have to touch any existing other table or the tests of them...

ivy-edp commented:
The pull request is now ready for merging:
- Integrated your input from your initial feedback regarding DocumentTable
- Adapted design and usability to other tables (CustomFieldTable, StartCustomFieldTable, MappingTree, ParameterTable)
- Added new Row: SelectableValidationRow (Combination of ValidationRow, MessageRow, SelectRow)
- Adjusted unit tests (UT) and integration tests (IT) for these tables

Open To-Dos:
- MailAttachmentPart, ConditionTable, PropertyTable, TableSort have been adjusted to eliminate UI bugs but still need further refinement in another PR. I have also skipped the tests (UT/IT) for now.
- I actually want to create a generic table component or outsource some parts, as I feel that I have repeated code in many places.

![grafik](https://github.com/axonivy/inscription-client/assets/141223521/fb676dd3-a154-420c-8116-ad4e831eba01)
![grafik](https://github.com/axonivy/inscription-client/assets/141223521/0c311855-4452-4c9f-a942-f87e2cd97990)
![grafik](https://github.com/axonivy/inscription-client/assets/141223521/b0bc8c0e-a545-4aef-839b-36b4949f88ef)
![grafik](https://github.com/axonivy/inscription-client/assets/141223521/0a2ef825-cd42-4740-aee0-904945ccc73f)
![grafik](https://github.com/axonivy/inscription-client/assets/141223521/b05a648c-adcd-47c6-b4b2-eaefdcedecd0)

ivy-lli commented:
First thing I already see from the screenshot, in the mapping tree the intention is no longer working.

ivy-lli commented on packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx:
```diff
@@ -16,15 +17,53 @@ const useValidationRow = (rowPathSuffix: string | number) => {
   );
 };
 
+export const SelectableValidationRow = <TData extends object>({
```
But you can do this also later in another PR.

ivy-lli commented on packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx:
```diff
@@ -1,122 +1,162 @@
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.value !== '');
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
+      <Table>
+        <thead>
+          {table.getHeaderGroups().map(headerGroup => (
+            <tr
+              key={headerGroup.id}
+              onClick={() => {
+                setRowSelection({});
+              }}
+              onDoubleClick={() => {
+                headerGroup.headers.forEach(header => {
+                  header.column.resetSize();
+                });
+              }}
+            >
```
reuseable component

ivy-lli commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
+            action: () => action({ name: table.getRowModel().rowsById[Object.keys(rowSelection)[0]].original.name, type })
+          },
+
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
not on the left side?

ivy-lli commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' ||
+            updatedData[updatedData.length - 1].type !== 'STRING' ||
+            updatedData[updatedData.length - 1].value !== '') &&
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.type !== 'STRING' || obj.value !== '');
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
+          ))}
+        </tbody>
+      </Table>
+      <TableAddRow colSpan={4} addRow={addRow} />
```
is this here always visible?

ivy-lli commented on packages/editor/src/components/parts/common/customfield/StartCustomFieldPart.tsx:
```diff
@@ -9,9 +9,9 @@ type StartCustomFieldPartProps = {
 const StartCustomFieldPart = ({ customFields, updateCustomFields }: StartCustomFieldPartProps) => {
   return (
-    <PathCollapsible label='Custom Fields' defaultOpen={customFields.length > 0} path='customFields'>
+    <PathContext path='customFields'>
```
I think you can move this context inside of the customfieldtable? and use there a pathcollapsible?

ivy-lli commented on packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:
```diff
@@ -1,108 +1,169 @@
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' ||
+            updatedData[updatedData.length - 1].type !== 'String' ||
+            updatedData[updatedData.length - 1].desc !== '') &&
```
why not use the emptyobject?

ivy-lli commented on packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:
```diff
@@ -1,108 +1,169 @@
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.type !== 'String' || obj.desc !== '');
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:
```diff
@@ -1,108 +1,169 @@
+  const tableActions =
+    table.getSelectedRowModel().rows.length > 0
+      ? [
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
isn't this on the left side of the row?

ivy-lli commented on packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:
```diff
@@ -1,108 +1,169 @@
-    </Table>
+          </tbody>
+        </Table>
+        {data.filter(obj => obj.name === '' && obj.type === 'String' && obj.desc === '').length === 0 && (
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/mail/MailAttachmentPart.tsx:
```diff
@@ -67,32 +67,32 @@ const MailAttachmentTable = () => {
+          ))}
+        </tbody>
+      </Table>
+      <TableAddRow colSpan={4} addRow={addRow} />
```
always visible?

ivy-lli commented on packages/editor/src/components/parts/name/document/DocumentTable.tsx:
```diff
@@ -1,113 +1,167 @@
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' || updatedData[updatedData.length - 1].url !== '') &&
```
why not use the empty object?

ivy-lli commented on packages/editor/src/components/parts/name/document/DocumentTable.tsx:
```diff
@@ -1,113 +1,167 @@
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.url !== '');
```
why not use the empty ebject?

ivy-lli commented on packages/editor/src/components/parts/name/document/DocumentTable.tsx:
```diff
@@ -1,113 +1,167 @@
+            icon: IvyIcons.GoToSource,
+            action: () => action(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].original.url)
+          },
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
isn't this on the left side?

ivy-lli commented on packages/editor/src/components/parts/name/document/DocumentTable.tsx:
```diff
@@ -1,113 +1,167 @@
+          ))}
+        </tbody>
+      </Table>
+      {data.filter(obj => obj.name === '' && obj.url === '').length === 0 && <TableAddRow colSpan={3} addRow={addRow} />}
```
why not use the empty object?

ivy-lli commented:
> First thing I already see from the screenshot, in the mapping tree the intention is no longer working.

Broken since 11.2.1. on 11.2.0 it is still working

ivy-edp commented on packages/editor/src/components/widgets/table/row/MessageRow.tsx:
```diff
@@ -1,24 +1,26 @@
-      </tr>
-      {message && (
-        <tr className='row row-message'>
-          <td colSpan={100}>
```
Yes, I have to set it now because otherwise, the validation message completely disrupts the entire design 😅. However, I have made the field optional now, and if it is not set, I have set the colSpan to 2.

ivy-edp commented on packages/editor/src/test-utils/table-utils.tsx:
```diff
@@ -28,54 +28,44 @@ export namespace TableUtil {
   export function assertReadonly() {
     expect(screen.getByRole('button', { name: 'Add row' })).toBeDisabled();
-    expect(screen.getAllByRole('button', { name: 'Remove row' })[0]).toBeDisabled();
+    expect(screen.queryByRole('button', { name: 'Remove row' })).toBeNull();
```
Yes, that is correct. However, from what I've seen, it's not possible to select a row in readonly mode. Therefore, I made this comparison directly, but when i think about it, I could also just remove this check. @ivy-lli What do you think?
![grafik](https://github.com/axonivy/inscription-client/assets/141223521/b5fb67ea-702b-41b7-9be2-619ba6cb6137)

ivy-edp commented on integrations/standalone/tests/pageobjects/Table.ts:
```diff
@@ -84,7 +84,8 @@ export class Row {
   async remove() {
-    await this.locator.getByRole('button', { name: 'Remove row' }).click();
+    await this.locator.click();
+    await this.page.getByRole('button', { name: 'Remove row' }).click();
```
Yes, I have also considered that, but unfortunately, it doesn't work with the rowsLocator since this button is completely outside of the table. I wouldn't know how to make it more specific, as the button simply belongs to the collapsible-header div, and there can be multiple of them on a page.

ivy-lli replied:
Leave it as it is for now 👍

ivy-edp commented on packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:
```diff
@@ -176,8 +163,8 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
       </Table>
-      <TableAddRow colSpan={4} addRow={addRow} />
-    </Collapsible>
+      {data.filter(obj => obj === EMPTY_WFCUSTOMFIELD).length === 0 && <TableAddRow addRow={addRow} />}
```
So I am correct in assuming that I should use deepEqual instead?

ivy-lli replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality

ivy-edp commented on packages/editor/src/components/parts/rest/rest-request/rest-body/RestForm.tsx:
```diff
@@ -119,10 +109,8 @@ export const RestForm = () => {
-          <TableAddRow colSpan={4} addRow={addRow} />
-        </TableFooter>
       </Table>
+      <TableAddRow addRow={addRow} />
```
Yes, for RestForm, RestParameters, TableSort, MailAttachementPart, and PropertyTable, the 'Add' line is currently always visible, as I haven't really adapted them to the new UX of the tables yet.

ivy-edp commented on packages/editor/src/components/widgets/table/row/ReorderRow.css:
```diff
@@ -5,12 +5,12 @@
-  padding: var(--size-2);
+  .ivy-change-type {
+    padding: var(--size-2);
+  }
```
Oh, I see. Sorry. I will adjust the CSS like this:
```
.dnd-row-handle .ivy-change-type {
  display: flex;
  justify-content: flex-end;
  padding: var(--size-2);
}
```
Because if I were to work with the <i> tag, the 'x' icon of the error message would also be affected, and I deliberately want to only affect the change-type icon with the padding.

ivy-lli replied:
Yes of course this is the other solution :)

ivy-edp commented on packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx:
```diff
@@ -37,29 +37,10 @@ export const SelectableValidationRow = <TData extends object>({
-        </tr>
-      )}
+      </SelectRow>
+      {message && <MessageRow colSpan={colSpan ? colSpan : 2} message={message} />}
```
Oh, yes, you are right. I will revert the change and address this component in the next PR. Sorry.
